Debugging guide (#12308)
pulumi/docs: content/docs/using-pulumi/debugging/_index.md

{
    while (!Debugger.IsAttached)
    {
        await System.Threading.Tasks.Task.Delay(1000);
    }
    Debugger.Break();

    // ... Your normal Pulumi code follows here
});
```

Run `pulumi up` the usual way:

```bash
pulumi up
Previewing update (dev)

 Type                 Name          Plan   Info
 pulumi:pulumi:Stack  play-cs-dev          'dotnet build -nologo .' completed successfully
```

Navigate to your IDE and attach the debugger to the process with the name of your .NET project. For example, in VS Code, you can define the following `launch.json` file:

```json
{
    "version": "0.2.0",
    "configurations": [
        {
            "name": ".NET Core Attach",
            "type": "coreclr",
            "request": "attach"
        }
    ]
}
```

navigate to the `Run and Debug` menu on the left pane, and then click the Start Debugging (F5) button with the `.NET Core Attach` configuration. Search for the process by your project name and attach:

![Debug: Attach to .NET Process](./dotnet-code-attach.png)

Now you can step through the program and inspect variables. Once completed, resume the execution and let the Pulumi CLI complete.

## Debugging Go programs

> Prepend your Pulumi program with the code to pause and give yourself time to attach the debugger:

Reviewer: I think you should mention that the Go program should be complied with debug flags using `go build -gcflags all=-N -l`, and then use the `runtime.binary` field in the project file to launch the built executable.

Author: > Please mention that the Go program should be compiled with debug flags using go build -gcflags all=-N -l, then used as the binary in Pulumi.yaml. Otherwise breakpoints may not work as expected.

I did not need to change anything about the compilation flags, it just worked. Was it just luck?

Reviewer: I think it _almost_ works without flags, but can be frustrating. [Here's](https://www.jetbrains.com/help/go/attach-to-running-go-processes-with-debugger.html#step-1-build-the-application) an example of the routine guidance.

---
title_tag: "Debugging Pulumi Programs"

> meta_desc: "Guide to attach a debugger to Pulumi programs and step through the code."

Reviewer 2: ```suggestion
meta_desc: "Guide to attaching a debugger to Pulumi programs and stepping through the code."
```

title: Debugging
h1: Debugging Pulumi programs
meta_image: /images/docs/meta-images/docs-meta.png
menu:
    usingpulumi:
        identifier: debugging
        weight: 9
---

Because Pulumi uses general purpose programming languages to provision cloud resources, you can take advantage of native debugging tools to troubleshoot your infrastructure definitions.

The tools differ depending on the language you are using, but the process stays similar:

> 1. Start with an existing Pulumi project, that contains `Pulumi.yaml` and your program.

Reviewer 2: ```suggestion
1. Start with an existing Pulumi project that contains `Pulumi.yaml` and your program.
```

2. For some languages, you'll have to temporarily modify your program to pause until a debugger is attached.
3. Set a breakpoint in your program.
4. Run `pulumi up` or `pulumi preview` from a command line.
5. Wait until `pulumi` is paused waiting for a debugger to attach.
6. Attach the debugger to a running process of your language runtime.
7. Step through the program to find the problem.
8. Run the program to completion and let `pulumi` shut down all the processes.
9. Change your program and repeat the process if needed.

Subsequent sections outline the specifics for each language runtime.

## Debugging Node.js programs

Run the Pulumi CLI by passing the [`--inspect-brk`](https://nodejs.org/en/learn/getting-started/debugging#command-line-options) option to node:

```bash
NODE_OPTIONS='--inspect-brk' pulumi up
```

The stack execution will pause until a debugger is attached:

```bash
Previewing update (dev)

     Type                      Name           Plan       Info
 +   pulumi:pulumi:Stack       aws-ts-dev     create     For help, see: https://nodejs.org/en/docs/inspector
```

Set a breakpoint in your program.

> Then, attach your debugger to the process with the `node` runtime. For example, in VS Code, you can define the following `launch.json` file:

Author: Great callout! I changed all line to something like 

```
For example, in [VS Code](https://code.visualstudio.com/docs/editor/debugging), you can define
the following `.vscode/launch.json` file:
```

```json
{
    "version": "0.2.0",
    "configurations": [
        {
            "name": "Node.js: Attach",
            "type": "node",
            "request": "attach",
            "continueOnAttach": true,
            "skipFiles": [
                "${workspaceFolder}/node_modules/**/*.js",
                "${workspaceFolder}/lib/**/*.js",
                "<node_internals>/**/*.js"
            ]
        }
    ]
}
```

navigate to the `Run and Debug` menu on the left pane, and then click the Start Debugging (F5) button with the `Node.js: Attach` configuration.

> Depending on your setup, you may need to attach multiple times until the right node process is found. For example, you may be running Node.js policy packs that will hit the debugger as well.

Author: We launch multiple `node` processes during `pulumi up` execution, and only one of them runs the user's program. Unfortunately, `NODE_OPTIONS` applies to all of them, so all `node` processes are launched in debug mode. In my sample project, I had to click `Run and Debug` twice to get to the breakpoint in my program.

```

Set a breakpoint in your Pulumi code.

Navigate to your IDE and attach the debugger to the process with the `python` runtime. For example, in VS Code, you can define the following `launch.json` file:

```json
{
    "version": "0.2.0",
    "configurations": [
        {
            "name": "Python: Local Attach",
            "type": "debugpy",
            "request": "attach",
            "connect": {
                "host": "localhost",
                "port": 5678
            },
            "pathMappings": [
                {
                    "localRoot": "${workspaceFolder}",
                    "remoteRoot": "."
                }
            ],
            "justMyCode": true
        }
    ]
}
```

navigate to the `Run and Debug` menu on the left pane, and then click the Start Debugging (F5) button with the `Python: Local Attach` configuration.

Now you can step through the program and inspect variables. Once completed, resume the execution and let the Pulumi CLI complete.

## Debugging .NET programs

Prepend your Pulumi program with the code to wait for a debugger to attach:

```cs
using System.Diagnostics;

> return await Pulumi.Deployment.RunAsync(async () =>

Reviewer 2: Might be worth mentioning the `async` bit is necessary here, as without it, you'll get a very noisy compile error.

Author: Added a comment to this effect

{
    while (!Debugger.IsAttached)
    {
        await System.Threading.Tasks.Task.Delay(1000);
    }

>     Debugger.Break();

Reviewer 2: FWIW, I stumbled a bit on this one as I interpreted this line as "being" the breakpoint -- e.g., that I had to move this line to where I wanted to breakpoint to happen. Might be worth mentioning this is just an initial breakpoint and that others can be set within the IDE.

Author: Removed now and added a suggestion for the user to set a breakpoint, similar to all other languages.

    // ... Your normal Pulumi code follows here
});
```

Run `pulumi up` the usual way:

```bash
pulumi up
Previewing update (dev)

 Type                 Name          Plan   Info
 pulumi:pulumi:Stack  play-cs-dev          'dotnet build -nologo .' completed successfully
```

> Navigate to your IDE and attach the debugger to the process with the name of your .NET project. For example, in VS Code, you can define the following `launch.json` file:

Reviewer 2: Any reason not to mention setting a breakpoint here also, as we do in the other sections?

Author: Because `Debugger.Break();` will break for the user. However, maybe it makes sense to unify this with other sections, remove `Debugger.Break();` and ask the user to set a breakpoint. I will do that.

```json
{
    "version": "0.2.0",
    "configurations": [
        {
            "name": ".NET Core Attach",
            "type": "coreclr",
            "request": "attach"
        }
    ]
}
```

> navigate to the `Run and Debug` menu on the left pane, and then click the Start Debugging (F5) button with the `.NET Core Attach` configuration. Search for the process by your project name and attach:

Reviewer 2: ```suggestion
Navigate to the `Run and Debug` menu on the left pane, and then click the Start Debugging (F5) button with the `.NET Core Attach` configuration. Search for the process by your project name and attach:
```

```bash
#!/bin/bash
go build -o goapp -gcflags "all=-N -l" .
exec dlv --listen=:2345 --headless=true --api-version=2 exec ./goapp
```

Be sure to allow to execute this script:

```bash
chmod +x debug.sh
```

Now, change your `Pulumi.yaml` file to invoke the debug script during program compilation:

```yaml
runtime:
  name: go
  options:
    binary: ./debug.sh
```

Run `pulumi up` the usual way:

```bash
pulumi up
Previewing update (dev)

 Type                 Name      Plan       Info
 pulumi:pulumi:Stack  dev       create...  warning layer=rpc Listening for remote connections
```

Set a breakpoint in your program.

Navigate to your IDE and connect the debugger to the same port as specified in the script (2345 in our case).

For example, in VS Code, you can define the following `launch.json` file:

```json
{
    "version": "0.2.0",
    "configurations": [
        {
            "name": "Connect to server",
            "type": "go",
            "request": "attach",
            "mode": "remote",
            "remotePath": "${workspaceFolder}",
            "port": 2345,
            "host": "127.0.0.1"
        }
    ]
}
```

> navigate to the `Run and Debug` menu on the left pane, and then click the Start Debugging (F5) button with the `Connect to server` configuration.

Reviewer 2: ```suggestion
Navigate to the `Run and Debug` menu on the left pane, and then click the Start Debugging (F5) button with the `Connect to server` configuration.
```

Now you can step through the program and inspect variables. Once completed, resume the execution and let the Pulumi CLI complete.

## Debugging Java programs

This approach assumes your project uses Maven and you have Maven shell scripts on your path.

Configure the runtime options in `Pulumi.yaml` to use `mvnDebug`:

```yaml
runtime:
  name: java
  options:
    use-executor: "mvnDebug"
```

Set a breakpoint in your Pulumi code.

Navigate to your IDE and attach to the port 8000. For example, in VS Code, you can define the following `launch.json` file:

```json
{
    "version": "0.2.0",
    "configurations": [
        {
            "type": "java",
            "name": "Debug (Attach)",
            "projectName": "App",
            "request": "attach",
            "hostName": "localhost",
            "port": 8000
        }
    ]
}
```

> navigate to the `Run and Debug` menu on the left pane, and then click the Start Debugging (F5) button with the `Debug (Attach)` configuration.

Reviewer 3: ```suggestion
Navigate to the `Run and Debug` menu on the left pane, and then click the Start Debugging (F5) button with the `Debug (Attach)` configuration.
```

5. Wait until `pulumi` is paused waiting for a debugger to attach.
6. Attach the debugger to a running process of your language runtime.
7. Step through the program to find the problem.
8. Run the program to completion and let `pulumi` shut down all the processes.
9. Change your program and repeat the process if needed.

Subsequent sections outline the specifics for each language runtime.

## Debugging Node.js programs

Run the Pulumi CLI by passing the [`--inspect-brk`](https://nodejs.org/en/learn/getting-started/debugging#command-line-options) option to node:

```bash
NODE_OPTIONS='--inspect-brk' pulumi up
```

The stack execution will pause until a debugger is attached:

```plain
Previewing update (dev)

     Type                      Name           Plan       Info
 +   pulumi:pulumi:Stack       aws-ts-dev     create     For help, see: https://nodejs.org/en/docs/inspector
```

Open your Pulumi program in your IDE and set a breakpoint at the line where you want the debugger to stop.

> Then, attach your debugger to the process with the `node` runtime. For example, in [VS Code](https://code.visualstudio.com/docs/editor/debugging), you can define the following `.vscode/launch.json` file:

Reviewer 4: Should we recommend using the JavaScript Debug Terminal if using VSCode - there's zero additional config required.

Author: Added a mention of auto-attach and debug terminal
